File: backend/app/domain/flex_width.py
"""按最小可读宽度修正 row 的横向分配。

solver 只按 ratios 瓜分宽度、不认识文字，所以「左边一大段正文 + 右边四张卡」
这类树会把卡片压到几十 pt，中文一行只剩四五个字。这里在生成期把过窄的列补回来：
先从同行宽裕的兄弟借，借不够就把并排的卡片折成两行。

纯函数，只改 row 的 ratios 与过窄子节点的分组，不动块内容与纵向 grow。
必须跑在 flex_fit 之前：高度度量依赖的列宽得先定下来。
"""

from __future__ import annotations

from app.domain.block_style import merge_text_style, resolve_box
from app.domain.content import Block
from app.domain.flex_layout import PRESET_INSET_PT, FlexContainer, FlexLeaf, FlexNode, is_spacer
from app.domain.flex_normalize import normalize
from app.domain.geometry import SAFE_AREA_WIDTH_PT
from app.domain.text_metrics import BULLET_INDENT_PT
from app.domain.theme import Theme

# 一行至少排得下这么多个全角字。参考成稿的卡片正文约 14 字/行，
# 这里取更宽松的下限：只拦「窄到读不了」的情况，不去规定审美。
MIN_CHARS_PER_LINE = 9.0

# 拉丁字母按全角的这个比例折算，与 text_metrics 的估算系数一致
_LATIN_EM = 0.55

# 与 app.render.pptx / 前端 CardsView 对齐的卡片排版常量
_CARD_GAP_PT = 16.0
_CARD_PAD_PT = 12.0
# KPI 在两端都有一个不小于 14pt 的内边距
_KPI_PAD_PT = 14.0

_MIN_VISUAL_WIDTH_PT = 96.0
_MIN_TABLE_COLUMN_PT = 56.0
_MIN_LEAF_WIDTH_PT = 48.0

# 两列 row 的 ratios 会被 normalize 吸附到设计比值，最多挪动 RATIO_SNAP_TOLERANCE。
# 预留略大于该容差的余量，避免刚补够宽度又被吸附吃回去。
_SNAP_SLACK_RATIO = 0.026

_EPS = 0.5

# 折行补救最多再走一轮：一次不够说明宽度是真不够，交给溢出告警。
_MAX_PASSES = 2


def fit_row_widths(
    tree: FlexContainer,
    blocks: list[Block],
    *,
    theme: Theme,
) -> FlexContainer:
    """返回横向分配已满足最小可读宽度的新树。

    只在生成期调用：用户手动拖出来的窄列是他自己的选择，不该被改回去。
    """
    ctx = _WidthContext(block_map={block.id: block for block in blocks}, theme=theme)
    return normalize(_fix_container(tree, SAFE_AREA_WIDTH_PT, ctx))


class _WidthContext:
    def __init__(self, *, block_map: dict[str, Block], theme: Theme) -> None:
        self.block_map = block_map
        self.theme = theme

    def min_width_pt(self, node: FlexNode) -> float:
        """节点内容排得下所需的最小宽度。"""
        return _min_width_pt(node, self)


def _fix_container(node: FlexContainer, avail_w_pt: float, ctx: _WidthContext) -> FlexContainer:
    if not node.children:
        return node
    if node.type == "row":
        return _fix_row(node, avail_w_pt, ctx, pass_index=0)
    inner = _inset(node, avail_w_pt)
    return node.model_copy(
        update={"children": [_fix_child(child, inner, ctx) for child in node.children]}
    )


def _fix_child(child: FlexNode, avail_w_pt: float, ctx: _WidthContext) -> FlexNode:
    if isinstance(child, FlexContainer):
        return _fix_container(child, avail_w_pt, ctx)
    return child


def _fix_row(
    node: FlexContainer,
    avail_w_pt: float,
    ctx: _WidthContext,
    *,
    pass_index: int,
) -> FlexContainer:
    n = len(node.children)
    inner_total = max(avail_w_pt - node.gap_pt * (n - 1), 1.0)
    percents = _ratio_percents(node)
    allocs = [inner_total * percent / 100.0 for percent in percents]
    # preset 会在每个子区域内再收一圈边，最小宽度要连这圈一起要
    chrome = 2 * PRESET_INSET_PT[node.preset] if node.preset is not None else 0.0
    slack = inner_total * _SNAP_SLACK_RATIO if n == 2 else 0.0
    needs = [ctx.min_width_pt(child) + chrome + slack for child in node.children]

    if all(alloc + _EPS >= need for alloc, need in zip(allocs, needs, strict=True)):
        return _recurse_row(node, allocs, ctx)

    if sum(needs) <= inner_total:
        balanced = _rebalance(allocs, needs)
        return _recurse_row(_with_allocs(node, balanced), balanced, ctx)

    regrouped = _regroup(node, allocs, needs, ctx, pass_index=pass_index)
    if regrouped is None:
        return _recurse_row(node, allocs, ctx)
    if regrouped.type != "row":
        return _fix_container(regrouped, avail_w_pt, ctx)
    if pass_index + 1 >= _MAX_PASSES:
        return _recurse_row(regrouped, _allocs_of(regrouped, avail_w_pt), ctx)
    return _fix_row(regrouped, avail_w_pt, ctx, pass_index=pass_index + 1)


def _recurse_row(node: FlexContainer, allocs: list[float], ctx: _WidthContext) -> FlexContainer:
    inner = [_inset(node, alloc) for alloc in allocs]
    return node.model_copy(
        update={
            "children": [
                _fix_child(child, width, ctx)
                for child, width in zip(node.children, inner, strict=True)
            ]
        }
    )


def _rebalance(allocs: list[float], needs: list[float]) -> list[float]:
    """让过窄的列拿到 need，缺口按富余比例从宽裕的兄弟身上扣，总宽不变。"""
    surplus = [max(0.0, alloc - need) for alloc, need in zip(allocs, needs, strict=True)]
    pool = sum(surplus)
    deficit = sum(max(0.0, need - alloc) for alloc, need in zip(allocs, needs, strict=True))
    if pool <= 0 or deficit <= 0:
        return list(allocs)
    ratio = min(1.0, deficit / pool)
    return [
        max(alloc, need) if alloc < need else alloc - room * ratio
        for alloc, need, room in zip(allocs, needs, surplus, strict=True)
    ]


def _regroup(
    node: FlexContainer,
    allocs: list[float],
    needs: list[float],
    ctx: _WidthContext,
    *,
    pass_index: int,
) -> FlexContainer | None:
    """把并排挤不下的一串子节点折成多行，返回改造后的容器；无从下手时返回 None。"""
    run = _longest_needy_run(allocs, needs)
    if run is None:
        return None

    gap = node.gap_pt
    group_w = sum(allocs[index] for index in run) + gap * (len(run) - 1)
    unit = max(needs[index] for index in run)
    per_row = int((group_w + gap) // (unit + gap))
    per_row = max(1, min(per_row, len(run) - 1))
    chunks = [run[start : start + per_row] for start in range(0, len(run), per_row)]
    if len(chunks) <= 1:
        return None

    rows = [
        FlexContainer(
            type="row",
            id=f"{node.id}__wrap{pass_index}r{index}",
            children=[node.children[position] for position in chunk],
            gap_pt=gap,
            ratios=[100.0 / len(chunk)] * len(chunk),
            grow=1.0,
        )
        for index, chunk in enumerate(chunks)
    ]
    wrapper = FlexContainer(
        type="column",
        id=f"{node.id}__wrap{pass_index}",
        children=rows,
        gap_pt=gap,
        grow=max(node.children[position].grow for position in run),
    )

    head, tail = run[0], run[-1] + 1
    children = [*node.children[:head], wrapper, *node.children[tail:]]
    if len(children) == 1:
        # 整行都折进去了，行本身没有存在意义，直接由 wrapper 顶替
        return wrapper.model_copy(update={"grow": node.grow})

    percents = _ratio_percents(node)
    ratios = [
        *percents[:head],
        sum(percents[head:tail]),
        *percents[tail:],
    ]
    return node.model_copy(update={"children": children, "ratios": ratios})


def _longest_needy_run(allocs: list[float], needs: list[float]) -> list[int] | None:
    """找出最长的一段连续「排不下」的子节点；不足两个就没得折。"""
    best: list[int] = []
    current: list[int] = []
    for index, (alloc, need) in enumerate(zip(allocs, needs, strict=True)):
        if alloc + _EPS < need:
            current.append(index)
            if len(current) > len(best):
                best = list(current)
        else:
            current = []
    return best if len(best) >= 2 else None


def _with_allocs(node: FlexContainer, allocs: list[float]) -> FlexContainer:
    total = sum(allocs)
    if total <= 0:
        return node
    return node.model_copy(update={"ratios": [alloc / total * 100.0 for alloc in allocs]})


def _allocs_of(node: FlexContainer, avail_w_pt: float) -> list[float]:
    n = len(node.children)
    inner_total = max(avail_w_pt - node.gap_pt * (n - 1), 1.0)
    return [inner_total * percent / 100.0 for percent in _ratio_percents(node)]


def _ratio_percents(node: FlexContainer) -> list[float]:
    n = len(node.children)
    ratios = node.ratios
    if ratios is None or len(ratios) != n:
        return [100.0 / n] * n
    total = sum(ratios)
    if total <= 0:
        return [100.0 / n] * n
    return [value / total * 100.0 for value in ratios]


def _inset(node: FlexContainer, width_pt: float) -> float:
    if node.preset is None:
        return width_pt
    return max(width_pt - 2 * PRESET_INSET_PT[node.preset], 1.0)


def _min_width_pt(node: FlexNode, ctx: _WidthContext) -> float:
    if isinstance(node, FlexLeaf):
        return _leaf_min_width_pt(node, ctx)
    if not node.children or is_spacer(node):
        return 0.0

    chrome = 2 * PRESET_INSET_PT[node.preset] if node.preset is not None else 0.0
    mins = [ctx.min_width_pt(child) + chrome for child in node.children]
    if node.type == "row":
        return sum(mins) + node.gap_pt * (len(mins) - 1)
    return max(mins)


def _leaf_min_width_pt(leaf: FlexLeaf, ctx: _WidthContext) -> float:
    block = ctx.block_map.get(leaf.block_id)
    if block is None:
        return _MIN_LEAF_WIDTH_PT

    theme = ctx.theme
    box = resolve_box(theme, block.style)
    padding = 2 * box.padding_pt

    def size_of(name: str) -> float:
        return merge_text_style(theme, name, block.style).size_pt

    match block.type:
        case "text":
            return _text_width_pt(block.text, size_of(leaf.text_style or "body")) + padding
        case "bullets":
            size_pt = size_of(leaf.text_style or "bullet")
            widest = max((_text_width_pt(item, size_pt) for item in block.items), default=0.0)
            return widest + BULLET_INDENT_PT + padding
        case "callout":
            # 渲染器按 variant 固定取样式，不看叶子的 text_style
            size_pt = size_of("caption" if block.variant == "source" else "body")
            return _text_width_pt(block.text, size_pt) + 2 * _CARD_PAD_PT
        case "cards":
            size_pt = size_of("body")
            card = max(
                max(_text_width_pt(item.title, size_pt), _text_width_pt(item.desc, size_pt))
                for item in block.items
            )
            n = len(block.items)
            return n * (card + 2 * _CARD_PAD_PT) + (n - 1) * _CARD_GAP_PT
        case "kpi":
            widest = max(
                _text_width_pt(text, size_of(name))
                for text, name in (
                    (block.value, "kpi_value"),
                    (block.label, "kpi_label"),
                    (block.note or "", "kpi_note"),
                )
            )
            return widest + 2 * max(box.padding_pt, _KPI_PAD_PT)
        case "table":
            return max(len(block.header), 1) * _MIN_TABLE_COLUMN_PT
        case "image" | "chart":
            return _MIN_VISUAL_WIDTH_PT

    return _MIN_LEAF_WIDTH_PT


def _text_width_pt(text: str, size_pt: float) -> float:
    """一行至少要排下 MIN_CHARS_PER_LINE 个字；文字本身更短就按它自己算。

    短标题不该被强行要求九个字的宽度，否则一行标题就能撑爆整个版面。
    """
    em = sum(_LATIN_EM if char.isascii() else 1.0 for char in text if not char.isspace())
    return min(MIN_CHARS_PER_LINE, em) * size_pt

File: java_backend/src/main/java/com/aippt/domain/flex/FlexWidth.java
package com.aippt.domain.flex;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.aippt.domain.Geometry;
import com.aippt.domain.TextMetrics;
import com.aippt.domain.Theme;
import com.aippt.domain.content.BlockStyle;
import com.aippt.domain.content.Blocks;

public final class FlexWidth {

    public static final double MIN_CHARS_PER_LINE = 9.0;
    private static final double LATIN_EM = 0.55;
    private static final double CARD_GAP_PT = 16.0;
    private static final double CARD_PAD_PT = 12.0;
    private static final double KPI_PAD_PT = 14.0;
    private static final double MIN_VISUAL_WIDTH_PT = 96.0;
    private static final double MIN_TABLE_COLUMN_PT = 56.0;
    private static final double MIN_LEAF_WIDTH_PT = 48.0;
    private static final double SNAP_SLACK_RATIO = 0.026;
    private static final double EPS = 0.5;
    private static final int MAX_PASSES = 2;

    private FlexWidth() {
    }

    public static FlexContainer fitRowWidths(FlexContainer tree, List<Map<String, Object>> blocks, Theme theme) {
        Map<String, Map<String, Object>> blockMap = new HashMap<>();
        for (Map<String, Object> block : blocks) {
            blockMap.put(Blocks.id(block), block);
        }
        return FlexNormalize.normalize(fixContainer(tree, Geometry.SAFE_AREA_WIDTH_PT, blockMap, theme));
    }

    private static FlexContainer fixContainer(
            FlexContainer node,
            double availW,
            Map<String, Map<String, Object>> blocks,
            Theme theme
    ) {
        if (node.children().isEmpty()) {
            return node;
        }
        if (node.isRow()) {
            return fixRow(node, availW, blocks, theme, 0);
        }
        double inner = inset(node, availW);
        List<FlexNode> children = new ArrayList<>();
        for (FlexNode child : node.children()) {
            children.add(fixChild(child, inner, blocks, theme));
        }
        return node.withChildren(children);
    }

    private static FlexNode fixChild(
            FlexNode child,
            double availW,
            Map<String, Map<String, Object>> blocks,
            Theme theme
    ) {
        if (child instanceof FlexContainer container) {
            return fixContainer(container, availW, blocks, theme);
        }
        return child;
    }

    private static FlexContainer fixRow(
            FlexContainer node,
            double availW,
            Map<String, Map<String, Object>> blocks,
            Theme theme,
            int pass
    ) {
        int n = node.children().size();
        double innerTotal = Math.max(availW - node.gapPt() * (n - 1), 1.0);
        List<Double> percents = ratioPercents(node);
        List<Double> allocs = percents.stream().map(p -> innerTotal * p / 100.0).toList();
        double chrome = node.preset() == null ? 0 : 2 * FlexSolve.PRESET_INSET_PT.getOrDefault(node.preset(), 0.0);
        double slack = n == 2 ? innerTotal * SNAP_SLACK_RATIO : 0;
        List<Double> needs = new ArrayList<>();
        for (FlexNode child : node.children()) {
            needs.add(minWidthPt(child, blocks, theme) + chrome + slack);
        }
        boolean ok = true;
        for (int i = 0; i < n; i++) {
            if (allocs.get(i) + EPS < needs.get(i)) {
                ok = false;
                break;
            }
        }
        if (ok) {
            return recurseRow(node, allocs, blocks, theme);
        }
        double needSum = needs.stream().mapToDouble(Double::doubleValue).sum();
        if (needSum <= innerTotal) {
            List<Double> balanced = rebalance(allocs, needs);
            return recurseRow(withAllocs(node, balanced), balanced, blocks, theme);
        }
        FlexContainer regrouped = regroup(node, allocs, needs, pass);
        if (regrouped == null) {
            return recurseRow(node, allocs, blocks, theme);
        }
        if (!regrouped.isRow()) {
            return fixContainer(regrouped, availW, blocks, theme);
        }
        if (pass + 1 >= MAX_PASSES) {
            return recurseRow(regrouped, allocsOf(regrouped, availW), blocks, theme);
        }
        return fixRow(regrouped, availW, blocks, theme, pass + 1);
    }

    private static FlexContainer recurseRow(
            FlexContainer node,
            List<Double> allocs,
            Map<String, Map<String, Object>> blocks,
            Theme theme
    ) {
        List<FlexNode> children = new ArrayList<>();
        for (int i = 0; i < node.children().size(); i++) {
            children.add(fixChild(node.children().get(i), inset(node, allocs.get(i)), blocks, theme));
        }
        return node.withChildren(children);
    }

    private static List<Double> rebalance(List<Double> allocs, List<Double> needs) {
        List<Double> surplus = new ArrayList<>();
        double pool = 0;
        double deficit = 0;
        for (int i = 0; i < allocs.size(); i++) {
            double room = Math.max(0, allocs.get(i) - needs.get(i));
            surplus.add(room);
            pool += room;
            deficit += Math.max(0, needs.get(i) - allocs.get(i));
        }
        if (pool <= 0 || deficit <= 0) {
            return new ArrayList<>(allocs);
        }
        double ratio = Math.min(1.0, deficit / pool);
        List<Double> result = new ArrayList<>();
        for (int i = 0; i < allocs.size(); i++) {
            result.add(allocs.get(i) < needs.get(i)
                    ? Math.max(allocs.get(i), needs.get(i))
                    : allocs.get(i) - surplus.get(i) * ratio);
        }
        return result;
    }

    private static FlexContainer regroup(FlexContainer node, List<Double> allocs, List<Double> needs, int pass) {
        List<Integer> run = longestNeedyRun(allocs, needs);
        if (run == null) {
            return null;
        }
        double gap = node.gapPt();
        double groupW = 0;
        for (int index : run) {
            groupW += allocs.get(index);
        }
        groupW += gap * (run.size() - 1);
        double unit = run.stream().mapToDouble(needs::get).max().orElse(1);
        int perRow = (int) ((groupW + gap) / (unit + gap));
        perRow = Math.max(1, Math.min(perRow, run.size() - 1));
        List<List<Integer>> chunks = new ArrayList<>();
        for (int start = 0; start < run.size(); start += perRow) {
            chunks.add(run.subList(start, Math.min(start + perRow, run.size())));
        }
        if (chunks.size() <= 1) {
            return null;
        }
        List<FlexNode> rows = new ArrayList<>();
        for (int index = 0; index < chunks.size(); index++) {
            List<Integer> chunk = chunks.get(index);
            List<FlexNode> kids = new ArrayList<>();
            for (int position : chunk) {
                kids.add(node.children().get(position));
            }
            rows.add(new FlexContainer(
                    "row",
                    node.id() + "__wrap" + pass + "r" + index,
                    kids,
                    gap,
                    equalRatios(chunk.size()),
                    null,
                    1.0
            ));
        }
        double grow = run.stream().mapToDouble(i -> node.children().get(i).grow()).max().orElse(1);
        FlexContainer wrapper = new FlexContainer("column", node.id() + "__wrap" + pass, rows, gap, null, null, grow);
        int head = run.get(0);
        int tail = run.get(run.size() - 1) + 1;
        List<FlexNode> children = new ArrayList<>();
        children.addAll(node.children().subList(0, head));
        children.add(wrapper);
        children.addAll(node.children().subList(tail, node.children().size()));
        if (children.size() == 1) {
            return wrapper.withGrow(node.grow());
        }
        List<Double> percents = ratioPercents(node);
        List<Double> ratios = new ArrayList<>(percents.subList(0, head));
        ratios.add(percents.subList(head, tail).stream().mapToDouble(Double::doubleValue).sum());
        ratios.addAll(percents.subList(tail, percents.size()));
        return node.withChildrenAndRatios(children, ratios);
    }

    private static List<Integer> longestNeedyRun(List<Double> allocs, List<Double> needs) {
        List<Integer> best = List.of();
        List<Integer> current = new ArrayList<>();
        for (int i = 0; i < allocs.size(); i++) {
            if (allocs.get(i) + EPS < needs.get(i)) {
                current.add(i);
                if (current.size() > best.size()) {
                    best = new ArrayList<>(current);
                }
            } else {
                current = new ArrayList<>();
            }
        }
        return best.size() >= 2 ? best : null;
    }

    private static FlexContainer withAllocs(FlexContainer node, List<Double> allocs) {
        double total = allocs.stream().mapToDouble(Double::doubleValue).sum();
        if (total <= 0) {
            return node;
        }
        return node.withChildrenAndRatios(node.children(), allocs.stream().map(a -> a / total * 100.0).toList());
    }

    private static List<Double> allocsOf(FlexContainer node, double availW) {
        int n = node.children().size();
        double innerTotal = Math.max(availW - node.gapPt() * (n - 1), 1.0);
        return ratioPercents(node).stream().map(p -> innerTotal * p / 100.0).toList();
    }

    private static List<Double> ratioPercents(FlexContainer node) {
        int n = node.children().size();
        List<Double> ratios = node.ratios();
        if (ratios == null || ratios.size() != n) {
            return equalRatios(n);
        }
        double total = ratios.stream().mapToDouble(Double::doubleValue).sum();
        if (total <= 0) {
            return equalRatios(n);
        }
        return ratios.stream().map(v -> v / total * 100.0).toList();
    }

    private static List<Double> equalRatios(int n) {
        List<Double> values = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            values.add(100.0 / n);
        }
        return values;
    }

    private static double inset(FlexContainer node, double widthPt) {
        if (node.preset() == null) {
            return widthPt;
        }
        return Math.max(widthPt - 2 * FlexSolve.PRESET_INSET_PT.getOrDefault(node.preset(), 0.0), 1.0);
    }

    private static double minWidthPt(FlexNode node, Map<String, Map<String, Object>> blocks, Theme theme) {
        if (node instanceof FlexLeaf leaf) {
            return leafMinWidth(leaf, blocks, theme);
        }
        FlexContainer container = (FlexContainer) node;
        if (container.children().isEmpty() || container.isSpacer()) {
            return 0;
        }
        double chrome = container.preset() == null ? 0 : 2 * FlexSolve.PRESET_INSET_PT.getOrDefault(container.preset(), 0.0);
        List<Double> mins = new ArrayList<>();
        for (FlexNode child : container.children()) {
            mins.add(minWidthPt(child, blocks, theme) + chrome);
        }
        if (container.isRow()) {
            return mins.stream().mapToDouble(Double::doubleValue).sum() + container.gapPt() * (mins.size() - 1);
        }
        return mins.stream().mapToDouble(Double::doubleValue).max().orElse(0);
    }

    @SuppressWarnings("unchecked")
    private static double leafMinWidth(FlexLeaf leaf, Map<String, Map<String, Object>> blocks, Theme theme) {
        Map<String, Object> block = blocks.get(leaf.blockId());
        if (block == null) {
            return MIN_LEAF_WIDTH_PT;
        }
        BlockStyle boxStyle = Blocks.styleOf(block);
        BlockStyle.ResolvedBox box = BlockStyle.resolveBox(theme, boxStyle);
        double padding = 2 * box.paddingPt();
        java.util.function.Function<String, Double> sizeOf = name ->
                BlockStyle.mergeTextStyle(theme, name, boxStyle).sizePt();
        return switch (Blocks.type(block)) {
            case "text" -> textWidth(Blocks.text(block), sizeOf.apply(leaf.textStyle() == null ? "body" : leaf.textStyle())) + padding;
            case "bullets" -> {
                double sizePt = sizeOf.apply(leaf.textStyle() == null ? "bullet" : leaf.textStyle());
                double widest = Blocks.items(block).stream().mapToDouble(item -> textWidth(item, sizePt)).max().orElse(0);
                yield widest + TextMetrics.BULLET_INDENT_PT + padding;
            }
            case "callout" -> {
                String variant = String.valueOf(block.getOrDefault("variant", "note"));
                double sizePt = sizeOf.apply("source".equals(variant) ? "caption" : "body");
                yield textWidth(Blocks.text(block), sizePt) + 2 * CARD_PAD_PT;
            }
            case "cards" -> {
                double sizePt = sizeOf.apply("body");
                Object items = block.get("items");
                double card = MIN_LEAF_WIDTH_PT;
                int n = 0;
                if (items instanceof List<?> list) {
                    n = list.size();
                    for (Object item : list) {
                        if (item instanceof Map<?, ?> map) {
                            card = Math.max(card, Math.max(
                                    textWidth(String.valueOf(map.get("title") == null ? "" : map.get("title")), sizePt),
                                    textWidth(String.valueOf(map.get("desc") == null ? "" : map.get("desc")), sizePt)
                            ));
                        }
                    }
                }
                yield n * (card + 2 * CARD_PAD_PT) + Math.max(0, n - 1) * CARD_GAP_PT;
            }
            case "kpi" -> {
                double widest = Math.max(
                        textWidth(String.valueOf(block.getOrDefault("value", "")), sizeOf.apply("kpi_value")),
                        Math.max(
                                textWidth(String.valueOf(block.getOrDefault("label", "")), sizeOf.apply("kpi_label")),
                                textWidth(String.valueOf(block.getOrDefault("note", "")), sizeOf.apply("kpi_note"))
                        )
                );
                yield widest + 2 * Math.max(box.paddingPt(), KPI_PAD_PT);
            }
            case "table" -> {
                Object header = block.get("header");
                int cols = header instanceof List<?> list ? Math.max(list.size(), 1) : 1;
                yield cols * MIN_TABLE_COLUMN_PT;
            }
            case "image", "chart" -> MIN_VISUAL_WIDTH_PT;
            default -> MIN_LEAF_WIDTH_PT;
        };
    }

    private static double textWidth(String text, double sizePt) {
        double em = 0;
        if (text != null) {
            for (int i = 0; i < text.length(); i++) {
                char ch = text.charAt(i);
                if (!Character.isWhitespace(ch)) {
                    em += ch < 128 ? LATIN_EM : 1.0;
                }
            }
        }
        return Math.min(MIN_CHARS_PER_LINE, em) * sizePt;
    }
}
